Ex2/Ex2.java

```diff
@@ -0,0 +1,209 @@
+			System.out.println("Node is not initialized yet.");
+			return;
+		}
+		
+		if (!tree.exists(targetNode)) {
+			System.out.println("No such node in the tree.");
+			return;
+		}
+		
+		System.out.println("Ancestors list of key " + targetNode + ":");
+		
+		if (tree.getRoot().getData() == targetNode.getData()) {
+			System.out.println("No ancestors! you're at the root!");
+			return;
+		}
+		
+		printAncestors(tree.getRoot(), targetNode);
+		System.out.println("|");
+		System.out.println();
+	}
+	
+	/**
+	 * A recursive function that outputs the ancestors of a given node.
+	 * The function iterates on the "next" node until it reaches the desired value
+	 * and then retrace its ancestors in the correct order
+	 * @param nextNode - a TreeNode to iterate on
+	 * @param targetNode - the desired node to find
+	 */
+	private static void printAncestors(TreeNode<?> nextNode, TreeNode<?> targetNode) {
+		if (nextNode.getData() == targetNode.getData()) {
+			System.out.print("Key - ");
+		}
+		
+		boolean hasRight = (nextNode.getRight() != null);
+		boolean hasLeft = (nextNode.getLeft() != null);
+		
+		if (hasRight && nextNode.getRight().exists(targetNode)) {
+			printAncestors(nextNode.getRight(), targetNode);
+			System.out.print(nextNode.getData() + " - ");
+			return;
+		}
+		
+		if (hasLeft && nextNode.getLeft().exists(targetNode)) {
+			printAncestors(nextNode.getLeft(), targetNode);
+			System.out.print(nextNode.getData() + " - " );
+		}	
+	}
+	
+	/**
+	 * This method return the lowest common ancestor node of two given nodes.
+	 * In case one of them is an ancestor of the other, it'll return the first one it encounter (oldest)
+	 * @param tree - the tree to operate on
+	 * @param node1 first node to check
+	 * @param node2 second node to check
+	 * @return the lowest common ancestor of node1 and node2
+	 */
+	public static TreeNode<?> lowestCommonAncestor(BinaryTree<?> tree, TreeNode<?> node1, TreeNode<?> node2) {
+				
+		if (tree == null || node1 == null || node2 == null) {
+			System.out.println("One of the parameters is not initialized correctly.");
+			return null;
+		}
+		
+		if (!tree.exists(node1) || !tree.exists(node2)) {
+			System.out.println("At least one of the nodes is not in the tree.");
+			return null;
+		}
+		
+		System.out.println("Lowest common ancestors of " + node1.toString() + " and " + node2.toString() +":");
+		return lowestCommonAncestor(tree.getRoot(), node1, node2);
+	}
+	
+	/**
+	 * This is a recursive function that checks every time if a candidate node
+	 * is the lowest common ancestor of node1 and node2.
+	 * @param candidate - Iterative node to check if lowest common ancestor of node1 and node2
+	 * @param node1 first node to check
+	 * @param node2 second node to check
+	 * @return The first Node which node1 and node2 are not located in its same subtree (left or right)
+	 */
+	private static TreeNode<?> lowestCommonAncestor(TreeNode<?> candidate, TreeNode<?> node1, TreeNode<?> node2) {
```

Comment:
Again, think about the efficiency of this algorithm. I think it descends down the tree multiple times. Can you do better?